pkg/controller/volume/attachdetach/metrics/metrics.go

```diff
@@ -0,0 +1,161 @@
+	podLister       corelisters.PodLister
+	pvLister        corelisters.PersistentVolumeLister
+	volumePluginMgr *volume.VolumePluginMgr
+}
+
+// nodePVCCount contains map of {"nodeName": {"pluginName": pvc_count }}
+// For example :
+//     node 172.168.1.100.ec2.internal has 10 EBS and 3 glusterfs PVC in use
+//     {"172.168.1.100.ec2.internal": {"aws-ebs": 10, "glusterfs": 3}}
+type nodePVCCount map[types.NodeName]map[string]int
+
+// Register registers pvc's in-use metrics
+func Register(pvcLister corelisters.PersistentVolumeClaimLister,
+	pvLister corelisters.PersistentVolumeLister,
+	podLister corelisters.PodLister,
+	pluginMgr *volume.VolumePluginMgr) {
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(newPVCInUseCollector(pvcLister, podLister, pvLister, pluginMgr))
+	})
+
+}
+
+func (pvcInUse nodePVCCount) add(nodeName types.NodeName, pluginName string) {
+	nodeCount, ok := pvcInUse[nodeName]
+	if !ok {
+		nodeCount = map[string]int{}
+	}
+	nodeCount[pluginName]++
+	pvcInUse[nodeName] = nodeCount
+}
+
+func newPVCInUseCollector(
+	pvcLister corelisters.PersistentVolumeClaimLister,
+	podLister corelisters.PodLister,
+	pvLister corelisters.PersistentVolumeLister,
+	pluginMgr *volume.VolumePluginMgr) *pvcInUseCollector {
+	return &pvcInUseCollector{pvcLister, podLister, pvLister, pluginMgr}
+}
+
+func (collector *pvcInUseCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- inUsePVCMetricDesc
+}
+
+func (collector *pvcInUseCollector) Collect(ch chan<- prometheus.Metric) {
+	nodePVCMap := collector.getPVCUseByNode()
+	for nodeName, pluginCount := range nodePVCMap {
+		for pluginName, count := range pluginCount {
+			metric, err := prometheus.NewConstMetric(inUsePVCMetricDesc,
+				prometheus.GaugeValue,
+				float64(count),
+				string(nodeName),
+				pluginName)
+			if err != nil {
+				glog.Warningf("Failed to create metric : %v", err)
+			}
+			ch <- metric
+		}
+	}
+}
+
+func (collector *pvcInUseCollector) getPVCUseByNode() nodePVCCount {
+	pods, err := collector.podLister.List(labels.Everything())
+	if err != nil {
+		glog.Errorf("Error getting pod list")
+		return nil
+	}
+
+	nodePVCMap := make(nodePVCCount)
+	for _, pod := range pods {
+		if len(pod.Spec.Volumes) <= 0 {
+			continue
+		}
+
+		nodeName := types.NodeName(pod.Spec.NodeName)
+		if nodeName == "" {
+			continue
+		}
+		for _, podVolume := range pod.Spec.Volumes {
+			volumeSpec, err := collector.CreateVolumeSpec(podVolume, pod.Namespace)
+			if err != nil {
+				continue
+			}
+			volumePlugin, err := collector.volumePluginMgr.FindPluginBySpec(volumeSpec)
+			if err != nil {
+				glog.V(3).Infof("Error finding volume plugin for : %v", volumeSpec)
+				continue
+			}
+			nodePVCMap.add(nodeName, volumePlugin.GetPluginName())
+		}
+	}
+	return nodePVCMap
+}
+
+func (collector *pvcInUseCollector) CreateVolumeSpec(podVolume v1.Volume, namespace string) (*volume.Spec, error) {
+	pvcSource := podVolume.VolumeSource.PersistentVolumeClaim
```

Comment:
I think it would be useful to include also inline volumes in pods and not just PVCs. It would bring broader (and more correct) picture.

Reply:
hrm, I overlooked something. turns out A/D controller only initializes attachable plugins (and similarly pv controller only initializes provisionable plugins) and hence it is pretty hard to emit a metric for all volumes in use without initiaizing all volume plugins in control plane.

The alternatives are:
1. Emit these metrics from kubelet. But the downside of that is, any unresponsive node could cause metrics to be incorrect.
2. Somehow find a way of initializing all plugins in control plane.

Reply:
for now I have just renamed the metric and added support for inline attachable volumes too. But yet to think how to truly report ALL volumes... not just attachable types.